diff --git a/audio/audio_driver.c b/audio/audio_driver.c
index 10dcbc5b3..dd959f900 100644
--- a/audio/audio_driver.c
+++ b/audio/audio_driver.c
@@ -597,6 +597,9 @@ static bool audio_driver_flush(const int16_t *data, size_t samples)
 
    audio_driver_resampler->process(audio_driver_resampler_data, &src_data);
 
+   audio_mixer_load_wav("/home/squarepusher/nav.wav");
+   audio_mixer_mix(audio_driver_output_samples_buf, src_data.output_frames);
+
    output_data   = audio_driver_output_samples_buf;
    output_frames = (unsigned)src_data.output_frames;
 
