Index: RetroArch_fork/retroarch.cfg
===================================================================
--- RetroArch_fork.orig/retroarch.cfg	2014-08-19 18:54:11.284946000 -0300
+++ RetroArch_fork/retroarch.cfg	2014-08-19 19:05:16.496941000 -0300
@@ -25,10 +25,10 @@
 # libretro_path = "/path/to/libretro.so"
 
 # A directory for where to search for libretro core implementations.
-# libretro_directory =
+libretro_directory = "/usr/lib/libretro/"
 
 # A directory for where to search for libretro core information.
-# libretro_info_path =
+libretro_info_path = "/usr/share/libretro/info/"
 
 # Sets log level for libretro cores (GET_LOG_INTERFACE).
 # If a log level issued by a libretro core is below libretro_log_level, it is ignored.
@@ -61,7 +61,7 @@
 # system_directory =
 
 # Sets start directory for menu content browser.
-# rgui_browser_directory =
+rgui_browser_directory = "~/"
 
 # Content directory. Interacts with RETRO_ENVIRONMENT_GET_CONTENT_DIRECTORY.
 # Usually set by developers who bundle libretro/RetroArch apps to point to assets.
@@ -69,10 +69,10 @@
 
 # Assets directory. This location is queried by default when menu interfaces try to look for
 # loadable assets, etc.
-# assets_directory =
+assets_directory = "/usr/share/libretro/assets/"
 
 # Sets start directory for menu config browser.
-# rgui_config_directory =
+rgui_config_directory = "~/.config/retroarch/"
 
 # Show startup screen in menu.
 # Is automatically set to false when seen for the first time.
@@ -172,13 +172,13 @@
 # video_shader_enable = false
 
 # Defines a directory where shaders (Cg, CGP, GLSL) are kept for easy access.
-# video_shader_dir =
+video_shader_dir = "/usr/share/libretro/shaders/"
 
 # CPU-based video filter. Path to a dynamic library.
 # video_filter =
 
 # Defines a directory where CPU-based video filters are kept.
-# video_filter_dir =
+video_filter_dir = "/usr/lib/retroarch/filters/video/"
 
 # Path to a font used for rendering messages. This path must be defined to enable fonts.
 # Do note that the _full_ path of the font is necessary!
@@ -235,7 +235,7 @@
 # audio_dsp_plugin =
 
 # Directory where DSP plugins are kept.
-# audio_filter_dir =
+audio_filter_dir = "/usr/lib/retroarch/filters/audio/"
 
 # Will sync (block) on audio. Recommended.
 # audio_sync = true
@@ -258,7 +258,7 @@
 #### Overlay
 
 # Defines a directory where overlays are kept for easy access.
-# overlay_directory =
+overlay_directory = "/usr/share/libretro/overlays/"
 
 # Path to input overlay
 # input_overlay =
@@ -294,7 +294,7 @@
 # Input binds which are made explicit (input_playerN_*_btn/axis) will take priority over autoconfigs.
 # Autoconfigs can be created with retroarch-joyconfig, manually, or with a frontend.
 # Requires input_autodetect_enable to be enabled.
-# joypad_autoconfig_dir =
+joypad_autoconfig_dir = "/usr/share/libretro/autoconfig/"
 
 # Sets which libretro device is used for a player.
 # Devices are indentified with a number.
