Index: RetroArch/retroarch.cfg
===================================================================
--- RetroArch.orig/retroarch.cfg	2014-07-10 01:27:27.793365000 -0300
+++ RetroArch/retroarch.cfg	2014-07-10 21:20:14.212388475 -0300
@@ -25,7 +25,7 @@
 # libretro_path = "/path/to/libretro.so"
 
 # A directory for where to search for libretro core implementations.
-# libretro_directory =
+libretro_directory = "/usr/lib/libretro/"
 
 # Sets log level for libretro cores (GET_LOG_INTERFACE).
 # If a log level issued by a libretro core is below libretro_log_level, it is ignored.
@@ -58,7 +58,7 @@
 # system_directory =
 
 # Sets start directory for menu content browser.
-# rgui_browser_directory =
+rgui_browser_directory = "~/"
 
 # Content directory. Interacts with RETRO_ENVIRONMENT_GET_CONTENT_DIRECTORY.
 # Usually set by developers who bundle libretro/RetroArch apps to point to assets.
@@ -69,7 +69,7 @@
 # assets_directory =
 
 # Sets start directory for menu config browser.
-# rgui_config_directory =
+rgui_config_directory = "~/.config/retroarch/"
 
 # Show startup screen in menu.
 # Is automatically set to false when seen for the first time.
